File: java/src/main/java/com/fairsearch/fair/Fair.java
package com.fairsearch.fair;

import com.fairsearch.fair.lib.FailprobabilityCalculator;
import com.fairsearch.fair.lib.MTableFailProbPair;
import com.fairsearch.fair.lib.MTableGenerator;
import com.fairsearch.fair.lib.RecursiveNumericFailprobabilityCalculator;
import com.fairsearch.fair.utils.FairScoreDoc;
import org.apache.lucene.search.TopDocs;

/**
 * This class serves as a wrapper around the utilities we have created for FA*IR ranking
 */
public class Fair {

    /**
     * Creates an mtable using alpha unadjusted.
     * @param n           Total number of elements
     * @param p           The proportion of protected candidates in the top-k ranking
     * @param alpha       The significance level
     * @return            The generated mtable, which is a int[]
     */
    public static int[] createUnadjustedMTable(int n, double p, double alpha) {
        MTableGenerator generator = new MTableGenerator(n, p, alpha, false);
        return generator.getMTable();
    }

    /**
     * Creates an mtable using alpha adjusted.
     * @param n           Total number of elements
     * @param p           The proportion of protected candidates in the top-k ranking
     * @param alpha       The significance level
     * @return            The generated mtable, which is a int[]
     */
    public static int[] createAdjustedMTable(int n, double p, double alpha) {
        MTableGenerator generator = new MTableGenerator(n, p, alpha, true);
        return generator.getMTable();
    }

    /**
     * Computes the alpha adjusted for the given set of parameters
     * @param n           Total number of elements
     * @param p           The proportion of protected candidates in the top-k ranking
     * @param alpha       The significance level
     * @return            The adjusted alpha
     */
    public static double adjustAlpha(int n, double p, double alpha) {
        RecursiveNumericFailprobabilityCalculator adjuster = new RecursiveNumericFailprobabilityCalculator(n, p, alpha);
        MTableFailProbPair failProbPair = adjuster.adjustAlpha();
        return failProbPair.getAlpha();
    }

    /**
     * Computes analytically the probability that a ranking created with the simulator will fail to pass the mtable
     * @param n           Total number of elements
     * @param p           The proportion of protected candidates in the top-k ranking
     * @param alpha       The significance level
     * @return            The adjusted alpha
     */
    public static double computeFailureProbability(int n, double p, double alpha) {
        FailprobabilityCalculator calculator = new RecursiveNumericFailprobabilityCalculator(n, p, alpha);
        return calculator.calculateFailprobability(n, p, alpha);
    }

    /**
     * Checks if the ranking is fair for the given parameters
     * @param docs        The ranking to be checked
     * @param mtable      The mtable against to check
     * @return            Returns whether the rankings statisfies the mtable
     */
    public static boolean checkRankingMTable(TopDocs docs, int[] mtable) {
        int countProtected = 0;

        //if the mtable has more elements than there are in the top docs return false
        if(docs.scoreDocs.length + 1 < mtable.length)
            return false;

        //check number of protected element at each ranking
        for(int i=0; i < docs.scoreDocs.length; i++) {
            countProtected += ((FairScoreDoc)docs.scoreDocs[i]).isProtected ? 1 : 0;
            if(countProtected < mtable[i+1])
                return false;
        }
        return true;
    }

    /**
     * Checks if the ranking is fair for the given parameters
     * @param docs        The ranking to be checked
     * @param n           Total number of elements
     * @param p           The proportion of protected candidates in the top-k ranking
     * @param alpha       The significance level
     * @return            Returns a boolean which specifies whether the ranking is fair
     */
    public static boolean isFair(TopDocs docs, int n, double p, double alpha) {
        MTableGenerator generator = new MTableGenerator(n, p, alpha, true);
        return checkRankingMTable(docs, generator.getMTable());
    }

    /**
     * Main method doing not much
     * @param args
     */
    public static void main(String[] args) {
        System.out.println(System.currentTimeMillis());
    }
}

File: python/fairsearchcore/fair.py
# -*- coding: utf-8 -*-

"""
This module serves as a wrapper around the utilities we have created for FA*IR ranking
"""
from fairsearchcore.alpha_adjustment import AlphaAdjustment

class Fair:

    k = 10 # the total number of elements
    p = 0.2 # the proportion of protected candidates in the top-k ranking
    alpha = 0.1 # the significance level

    def __init__(self, k: int, p: float, alpha: float):
        # check the parameters first
        _validate_basic_parameters(k, p, alpha)

        # assign the parameters
        self.k = k
        self.p = p
        self.alpha = alpha

    def _create_mtable(self, alpha: int, adjust_alpha: float):
        """
        Creates an mtable by, if you want, pass your own alpha value(overridng the object's one)
        :param alpha:           The significance level
        :param adjust_alpha:    Boolean indicating whether the alpha be adjusted or not
        :return:
        """
        # check if passed alpha is ok
        _validate_alpha(alpha)

        # create the mtable
        mtable = None
        fc = AlphaAdjustment(k, self.__minProp, self.__alpha)
        return mtable

    def create_unadjusted_mtable(self):
        """
        Creates an mtable using alpha unadjusted
        :return:
        """
        pass

    def create_adjusted_mtable(self):
        """
        Creates an mtable using alpha adjusted
        :return:
        """
        pass

    def adjust_alpha(self):
        """
        Computes the alpha adjusted for the given set of parameters
        :return:
        """
        pass

    def compute_fail_probability(self):
        """
        Computes analytically the probability that a ranking created with the simulator will fail to pass the mtable
        :return:
        """

    def is_fair(self, ranking: list):
        """
        Checks if the ranking is fair for the given parameters
        :param ranking:     The ranking to be checked
        :return:
        """
        return check_ranking(ranking, self.create_adjusted_mtable())

def check_ranking(ranking:list, mtable: list):
    """
    Checks if the ranking is fair in respect to the mtable
    :param ranking:     The ranking to be checked
    :param mtable:      The mtable against to check
    :return:            Returns whether the rankings statisfies the mtable
    """
    count_protected = 0

    # if the mtable has a different number elements than there are in the top docs return false
    if len(ranking) != len(mtable):
        raise ValueError("Number of documents in ranking and mtable length are not the same!")

    # check number of protected element at each rank
    for i, element in ranking:
        count_protected += 1 if element == 1 else 0
        if count_protected < mtable[i]:
            return False
    return True

def _validate_basic_parameters(k, p, alpha):
    """
    Validates if k, p and alpha are in the required ranges
    :param k:           Total number of elements (above or equal to 10)
    :param p:           The proportion of protected candidates in the top-k ranking (between 0.02 and 0.98)
    :param alpha:       The significance level (between 0.01 and 0.15)
    """
    if k < 10:
        raise ValueError("Total number of elements `k` must be above or equal to 10")
    if p < 0.02 or p > 0.98:
        raise ValueError("The proportion of protected candidates `p` in the top-k ranking must between 0.02 and 0.98")

    _validate_alpha(alpha)

def _validate_alpha(alpha):
    if alpha < 0.01 or alpha > 0.15:
        raise ValueError("The significance level `alpha` must be between 0.01 and 0.15")
